lib/jsonapi/utils/underscore.ex

```diff
@@ -56,14 +60,44 @@ defmodule JSONAPI.Utils.Underscore do
   end
 
   def underscore(value) when is_map(value) do
-    Enum.into(value, %{}, &underscore/1)
+    Enum.into(value, %{}, &underscore(&1))
   end
 
   def underscore({key, value}) do
-    if is_map(value) do
-      {underscore(key), underscore(value)}
+    if underscore?(key) do
+      if is_map(value) do
+        {underscore(key), underscore(value)}
+      else
+        {underscore(key), value}
+      end
     else
       {underscore(key), value}
     end
   end
+
+  defp underscore?(key) do
+    config = Application.get_env(:jsonapi, :underscore_to_dash)
+    config_specifies_underscore?(config, key)
+  end
+
+  defp config_specifies_underscore?(true, _), do: true
+  defp config_specifies_underscore?(false, _), do: false
+
+  defp config_specifies_underscore?(config, key) when is_list(config) do
+    if Keyword.has_key?(config, :only) do
```

Comment:
Would this suffice?

```elixir
  defp config_specifies_underscore?(config, key) do
    config_specifies?(config, :only, key) or not config_specifies?(config, :except, key)
  end
```

Since you're guarding in `config_specifies?/3` and handling an error case, we don't need to check if the either `:only` or `:except` exists we can just call the functions.

Reply:
No, in fact that is how I originally wanted to write it, but it is not the same logic. If `:only` is specified, then we do not want to check `:except` at all, regardless of whether the `:only` specifies a key. Especially since `:except` is negated, which means that if neither `:only` nor `:except` is specified, then `not config_specifies?(config, :except, key)` will return true.

Reply:
Here are a couple of alternatives that keep the same logic, and remove  the indirection of `config_specifies/3`:

```elixir
# All in one if
defp config_specifies_underscore?(config, key) when is_list(config) do
  if Keyword.has_key?(config, :only) do
    key in Keyword.get(config, :only, [])
  else
    key not in Keyword.get(config, :except, [])
  end
end
```

```elixir
# Cond
defp config_specifies_underscore?(config, key) when is_list(config) do
  cond do
    Keyword.has_key?(config, :only) -> key in Keyword.get(config, :only, [])
    Keyword.has_key?(config, :except) -> key not in Keyword.get(config, :except, [])
    true -> false
  end
end
```

Is either of those alternatives preferable?

Reply:
second one looks good!